[DCP] Avoid multiple storage writer resets in async save

saumishr opened the pull request:
Summary: Avoid multiple storage writer resets in async save. Currently the reset gets called by the async_save method and then again in the save method. In the async path, async_save should only do the staging and the reset should only happen in the synchronous save path.

Test Plan:
```
buck test 'fbcode//mode/opt' //aiplatform/modelstore/experimental/DCP/tests:checkpoint_dist_client_test
```
https://www.internalfb.com/intern/testinfra/testrun/15199648841705052

Rollback Plan:

Differential Revision: D79230339




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

Skylion007 commented on torch/distributed/checkpoint/_async_process_executor.py:
```diff
@@ -172,6 +178,8 @@ def _execute_save(
         checkpoint_request_id: _CheckpointRequestIdentifier,
         storage_writer: Optional[StorageWriter] = None,
         planner: Optional[SavePlanner] = None,
+        no_dist: bool = False,
```
Starting to fear the Boolean trap here? Is use_collectives valid with no_dist True or should this actually be an enum?

Skylion007 commented on torch/distributed/checkpoint/filesystem.py:
```diff
@@ -737,8 +760,12 @@ def finish(self, metadata: Metadata, results: list[list[WriteResult]]) -> None:
+            if not self.use_collectives and self.rank is not None
+            else f"{_metadata_fn}.tmp"
+        )
+        tmp_path = cast(Path, self.fs.concat_path(self.path, tmp_filename))
```
Nit: should some of these casts be PathLike?

saumishr committed:
[DCP] Avoid multiple storage writer resets in async save (#159448)

Summary:

Avoid multiple storage writer resets in async save. Currently the reset gets called by the async_save method and then again in the save method. In the async path, async_save should only do the staging and the reset should only happen in the synchronous save path.

Test Plan:
```
buck test 'fbcode//mode/opt' //aiplatform/modelstore/experimental/DCP/tests:checkpoint_dist_client_test
```
https://www.internalfb.com/intern/testinfra/testrun/15199648841705052

Rollback Plan:

Reviewed By: meetv18, Saiteja64

Differential Revision: D79230339
